packages/middleware-allow-request-methods/lib/index.js

```diff
@@ -0,0 +1,112 @@
+const { logRecoverableError } = require('@dotcom-reliability-kit/log-error');
+
+/**
+ * @typedef {object} RequestMethodOptions
+ * @property {string[]} [allowedMethods] The HTTP methods that are allowed i.e. will not throw 405 errors.
+ * @property {string} [message] The error message text to use if a disallowed method is used.
+ * @property {import('@dotcom-reliability-kit/log-error').Logger} [logger] The logger to use for logging errors.
+ */
+
+/**
+ * @typedef {import('express').ErrorRequestHandler} ExpressErrorHandler
+ */
+
+/**
+ * Create a middleware function to return 405 (rather than 404) for disallowed request methods.
+ *
+ * @param {RequestMethodOptions} [options]
+ * @returns {ExpressErrorHandler}
+ */
+function allowRequestMethods(
+	options = { allowedMethods: [], message: 'Method Not Allowed' }
+) {
+	const normalisedAllowedRequestMethods = normaliseAllowedRequestMethods(
+		options.allowedMethods || []
+	);
+
+	return function allowRequestMethodsMiddleware(
+		error,
+		request,
+		response,
+		next
+	) {
+		// If headers are already sent, pass the error to the default Express error handler
+		if (response.headersSent) {
+			return next(error);
+		}
+
+		try {
+			// If the allowed methods array is empty, you can either allow all methods or reject everything
+			if (normalisedAllowedRequestMethods.length === 0) {
+				// TODO: Option 1: Allow all methods (no restriction) i.e. request proceeds as normal
+				return next();
+
+				// TODO: or Option 2: Reject all methods (405 for every request) i.e. block all requests when no methods are explicitly stated
+				// response.header('Allow', normalisedAllowedRequestMethods.join(', '));
+				// response.status(405).send(options.message);
+				// return next(new MethodNotAllowedError(options.message));
+			}
```

Comment:
Had a thought on the intention of this function and felt it could either be option 1 or option 2, but want to double check, as I initially wrote it assuming option 2 (i.e. I'm assuming if someone's using this package then they're using it with the intention of it's stated purpose and thus they'll explicitly state the allowed methods for that route, else why would they be requiring this package into their app 🤷🏽‍♀️ )

Reply:
Yep I agree with your gut here - option 1 should not be a use-case, if you don't want to protect an endpoint then don't use this middleware. I think we should go one further and throw a `TypeError` if no allowed methods are specified. We do similar option validation in other places, [e.g. here](https://github.com/Financial-Times/dotcom-reliability-kit/blob/d2058bfdc140b8144a080a32865b57de08bc8869/packages/middleware-log-errors/lib/index.js#L34-L38)